framework/src/tinyalsa/tinyalsa.c

```diff
@@ -532,50 +533,82 @@ int pcm_readi(struct pcm *pcm, void *data, unsigned int frame_count)
 	nbytes = pcm_frames_to_bytes(pcm, frame_count);
 
 #ifdef CONFIG_AUDIO_DRIVER_SPECIFIC_BUFFERS
-	if (nbytes < pcm->config.period_size) {
-		return -EINVAL;
-	}
 	bufdesc.numbytes = pcm->config.period_size;
 #else
-	if (nbytes < CONFIG_AUDIO_BUFFER_NUMBYTES) {
-		return -EINVAL;
-	}
 	bufdesc.numbytes = CONFIG_AUDIO_BUFFER_NUMBYTES;
 #endif
 #ifdef CONFIG_AUDIO_MULTI_SESSION
 	bufdesc.session = pcm->session;
 #endif
 
 	/* If device is not yet started, start now! */
-	if ((!pcm->running) && (pcm_start(pcm) < 0)) {
-		return -errno;
-	} else {
-		pcm->running = 1;
+	if (!pcm->running) {
+		pcm->enqueuedBytes = 0;
+		/* If the device is opened for read and Our buffers are not enqued. Enque them now. */
+#ifdef CONFIG_AUDIO_MULTI_SESSION
+		bufdesc.session = pcm->session;
+#endif
+#ifdef CONFIG_AUDIO_DRIVER_SPECIFIC_BUFFERS
+		bufdesc.numbytes = pcm->config.period_size;
+		for (pcm->bufPtr = 0; pcm->bufPtr < buf_info.nbuffers; pcm->bufPtr++)
+#else
+		bufdesc.numbytes = CONFIG_AUDIO_BUFFER_NUMBYTES;
+		for (pcm->bufPtr = 0; pcm->bufPtr < CONFIG_AUDIO_NUM_BUFFERS; pcm->bufPtr++)
+#endif
+		{
+			bufdesc.u.pBuffer = pcm->pBuffers[pcm->bufPtr];
+			if (nbytes <= pcm->enqueuedBytes) {
+				break;
+			} else if (nbytes - pcm->enqueuedBytes < bufdesc.numbytes) {
+				bufdesc.numbytes = nbytes - pcm->enqueuedBytes;
+			}
+			bufdesc.u.pBuffer->nmaxbytes = bufdesc.numbytes;
+			bufdesc.u.pBuffer->nbytes = 0;
+			bufdesc.u.pBuffer->curbyte = 0;
+			bufdesc.u.pBuffer->flags = 0;
+
+			if (ioctl(pcm->fd, AUDIOIOC_ENQUEUEBUFFER, (unsigned long)&bufdesc) < 0) {
+				return oops(pcm, errno, "AUDIOIOC_ENQUEUEBUFFER ioctl failed");
+			}
+			pcm->enqueuedBytes += bufdesc.numbytes;
+		}
+		if (pcm_start(pcm) < 0) {
+			return -errno;
+		}
 	}
 
-	/* Wait for deque message from kernel */
 	size = mq_receive(pcm->mq, (FAR char *)&msg, sizeof(msg), &prio);
 	if (size != sizeof(msg)) {
 		/* Interrupted by a signal? What to do? */
 		return oops(pcm, EINTR, "Interrupted while waiting for deque message from kernel");
 	}
 	if (msg.msgId == AUDIO_MSG_DEQUEUE) {
 		apb = (struct ap_buffer_s *)msg.u.pPtr;
-		/* Copy data to user buffer */
+		pcm->enqueuedBytes -= apb->nmaxbytes;
+		pcm->enqueuedBytes -= apb->nmaxbytes - apb->nbytes;
+
 		memcpy(data, apb->samp, apb->nbytes);
-		/* Enque buffer for next read opertion */
-		nbytes = apb->nbytes;
-		apb->nbytes = 0;
-		apb->curbyte = 0;
-		apb->flags = 0;
-		bufdesc.u.pBuffer = apb;
-		if (ioctl(pcm->fd, AUDIOIOC_ENQUEUEBUFFER, (unsigned long)&bufdesc) < 0) {
-			return oops(pcm, errno, "failed to enque buffer after read");
+		if (nbytes - apb->nbytes > pcm->enqueuedBytes) {
```

Comment:
If this condition fails, you will return from readi without enqueue-ing the buffer. 
During next call to readi, you will no longer have a pointer to this buffer. So, this buffer will get lost.